### PR TITLE
Output transpiled circuit in dt

### PR DESCRIPTION
**Pull request by lcapelluto**
### Summary

The circuit instruction `delay` will now be transpiled to units understood
by the target backend, i.e. units of dt. Thus, the following valid code
will execute as expected:

```
        qc = QuantumCircuit(1)
        qc.delay(100, [0], unit='us')
        qc = transpile(qc, backend)
        job = execute(qc, backend)
```

Previously, the backend would delay for 100 timesteps (each of duration dt)
rather than (100e-6 / dt) timesteps.

### Details and comments

Any reasons not to do this? Should this not be done for `optimization_level=0` for instance?

I'm convinced that the example code ought to delay for 100 microseconds, and not 100 timesteps, but I'm very open to alternative solutions. (Should probably be a proper pass, but a 35 line diff is preferable until we decide on the interface and behavior.) This didn't really matter before backends accepted the `delay` instruction natively, which they now do.

### CODE REVIEW

File: qiskit/compiler/transpile.py [hunk 1]
```diff
@@ -239,6 +240,11 @@ def callback_func(**kwargs):
 
     # Transpile circuits in parallel
     circuits = parallel_map(_transpile_circuit, list(zip(circuits, transpile_args)))
+    if not scheduling_method:
```
Comment (lcapelluto) [inline qiskit/compiler/transpile.py:243]: * should it be replaced with a transpiler pass?

**Comment by ewinston** [comment]
It looks like when the delay is not a multiple dt, it doesn't warn and doesn't display the rounded delay time. It seems like after transpilation it should at least do the later.

**Review by itoko — COMMENTED** [review]
I understand the issue is in the spec of `execute()` or `transpile(scheduling_method=None)`.
Because, if we transpile circuits with delays in seconds specifying `scheduling_method` and `backend` which provides `dt`, we have no problem, all the delays in output scheduled circuits have durations in dt.

Is the original issue there is no way to directly execute "partially scheduled circuit" (a normal circuit with delays)?
I'm not sure if we should allow the execution of circuits with delays without scheduling.
I think users of delays aware on timing of instructions, so I think we should distinguish a normal circuit with delays (timing aware) as a normal circuit (timing agnostic). 
We already have three executable objects: "normal circuit", "scheduled circuit" and "schedule".
I have some reservations on introducing one more in-between "normal circuit" and "scheduled circuit". I'm also concerning users have no means to know the actual "scheduled circuit" run on the backend if we allow the execution of "partially scheduled circuit".
For example, if given this circuit,
```
qc = QuantumCircuit(2)
qc.h(0)
qc.delay(100, 0)
qc.h(1)
qc.measure_all()
```
the timing of the Hadamard gate to qubit 1 is determined when scheduling (on the backend).

Or, is the original issue "there is no way to transpile and run circuits with delays at once"?
In this case,  we can think of simple solutions:
- Set scheduling_method="alap" behind the scene when executing a circuit with delays
- Ask users to specify scheduling_method to execute circuits with delays, and tell the scheduling_method argument in execute() to transpile().

Any thoughts? @lcapelluto @ajavadia

**Comment by lcapelluto** [comment]
In my opinion, `delay` can be treated exactly the same way that other supported gates are treated. This is more similar to getting the units wrong.

As an analog, we could image we have a backend that supports `RX(theta)` and `theta` is expected in radians, but the Qiskit interface accepts `theta` in units of degrees, ex. `RX(90, [0], unit='degrees')` (same syntax here as delay). If I add this operation to a circuit, transpile, and execute, it seems weird that the backend would execute `RX(90 radians)`.

Either way, I'm happy to follow @ajavadia's decision.

**Comment by itoko** [comment]
Thank you for elaborating on what is the issue. I think I understand your point now.
The point is who should convert the unit of delays. Currently it owes to users (when they don't specify `scheduling_method`) and you are suggesting it should be done within transpile().
I agree with you in that transpile() should convert the unit.
I'm even thinking transpile() should always schedule a circuit if it includes delays (timing-aware instructions), and the scheduling includes conversion of the unit.

In other words, my concern is in execute() rather than transpile().
In my opinion, backends should accept only the "fully scheduled circuits" for timing-aware circuits, so circuits with delays should be always scheduled within execute().

I'm happy to follow @ajavadia's decision, too.

**Comment by lcapelluto** [comment]
I think we mostly agree, except on the point of whether converting the delay units to whatever the backend expects should always be accompanied by deciding on the exact timing of every other operation

**Comment by ajavadia** [comment]
Good point, I think inserting delay means you are expressing a timing intent on the circuit, so I think that should result in a fully scheduled circuit (i.e. don't leave anything to the backend scheduler, insert delay everywhere to yield a scheduled circuit). I'm fine if we invoke the `ALAPScheduler` pass behind the scenes in this case (in all preset passmanagers). This pass should take care of the unit conversion.

Also @ewinston's point above is right, it should raise a warning somewhere if the units are not converted cleanly. But I thought we had this somewhere?

**Comment by lcapelluto** [comment]
Thank you for the discussion, @ajavadia. Updated!:)

File: qiskit/compiler/transpile.py [hunk 2]
```diff
@@ -617,9 +618,17 @@ def _parse_translation_method(translation_method, num_circuits):
...
-def _parse_scheduling_method(scheduling_method, num_circuits):
+def _parse_scheduling_method(scheduling_method, circuits):
+    """If there is a delay in any circuit, implicitly add a default scheduling method."""
+    if scheduling_method is None:
+        for circ in circuits:
+            for inst, _, _ in circ:
+                if isinstance(inst, Delay):
+                    scheduling_method = 'alap'
+                    break
+
```
Comment (itoko) [inline qiskit/compiler/transpile.py:629]:
```suggestion
            else:
                continue
            break
```
maybe better to break from outer loop too.

**Review by itoko — COMMENTED** [review]
LGTM. Just one minor comment.

File: qiskit/compiler/transpile.py (hunk 2 above)
Comment (lcapelluto) [reply qiskit/compiler/transpile.py:629]: I just double checked; these are equivalent

Comment (lcapelluto) [reply qiskit/compiler/transpile.py:628]: Lmk if you have a way to prove otherwise

Comment (lcapelluto) [reply qiskit/compiler/transpile.py:628]: You're right that it doesn't break from the outer loop though. The alternative is:

```
def _parse_scheduling_method(scheduling_method, circuits):
    """If there is a delay in any circuit, implicitly add a default scheduling method."""
    delay_found = False
    if scheduling_method is None:
        for circ in circuits:
            if delay_found:
                break
            for inst, _, _ in circ:
                if isinstance(inst, Delay):
                    scheduling_method = 'alap'
                    delay_found = True
                    break
```

Comment (lcapelluto) [reply qiskit/compiler/transpile.py:628]: Oh I just learned about this! Your suggestion works if it's unindented one level:
```
def _parse_scheduling_method(scheduling_method, circuits):
    """If there is a delay in any circuit, implicitly add a default scheduling method."""
    if scheduling_method is None:
        for circ in circuits:
            for inst, _, _ in circ:
                if isinstance(inst, Delay):
                    scheduling_method = 'alap'
                    break
            else:
                continue
            break
```

Comment (lcapelluto) [reply qiskit/compiler/transpile.py:628]: I think it might be more readable to do:
```
def _parse_scheduling_method(scheduling_method, circuits):
    """If there is a delay in any circuit, implicitly add a default scheduling method."""
    def has_delay(circuit):
        for inst, _, _ in circuit:
            if isinstance(inst, Delay):
                return True
        return False

    if scheduling_method is None:
        for circ in circuits:
            if has_delay(circ):
                scheduling_method = 'alap'
                break
```

Comment (lcapelluto) [reply qiskit/compiler/transpile.py:633]: Updated with the last one

File: test/python/circuit/test_scheduled_circuit.py [hunk 3]
```diff
@@ -164,13 +164,21 @@ def test_transpile_circuit_with_custom_instruction(self):
...
-    def test_transpile_delay_circuit_without_scheduling_method_as_normal_circuit(self):
+    def test_transpile_delay_circuit_without_scheduling_method(self):
         qc = QuantumCircuit(2)
         qc.h(0)
         qc.delay(500, 1)
         qc.cx(0, 1)
-        transpiled = transpile(qc)
-        self.assertEqual(transpiled.duration, None)
+        transpiled = transpile(qc, backend=self.backend_with_dt)
+        self.assertEqual(transpiled.duration, 2260)
```
Comment (lcapelluto) [inline test/python/circuit/test_scheduled_circuit.py:173]: Note the altered behavior here, requiring delays to trigger implicit scheduling is no longer backwards compatible @ajavadia @itoko

Comment (lcapelluto) [reply test/python/circuit/test_scheduled_circuit.py:173]: How do you want this handled?

Comment (itoko) [reply test/python/circuit/test_scheduled_circuit.py:173]: Ah, the original test case is renamed to `test_transpile_delay_circuit_without_scheduling_method_or_durs` but git missed it...

Anyways, I personally think it's ok to change the behavior in these cases because they are more like fixing a bug in a spec itself and we cannot fix them without losing backwards compatibility. Adding note (as you've already done so, thanks!) would be sufficient.

File: qiskit/compiler/transpile.py [hunk 4]
```diff
@@ -418,13 +419,17 @@ def _parse_transpile_args(circuits, backend,
...
     seed_transpiler = _parse_seed_transpiler(seed_transpiler, num_circuits)
     optimization_level = _parse_optimization_level(optimization_level, num_circuits)
     output_name = _parse_output_name(output_name, circuits)
     callback = _parse_callback(callback, num_circuits)
 
+    durations = _parse_instruction_durations(backend, instruction_durations, dt, circuits)
+    scheduling_method = _parse_scheduling_method(scheduling_method, circuits)
+    if scheduling_method and not durations:
+        raise TranspilerError("Transpiling a circuit with a scheduling method or with delay "
+                              "instructions requires a backend or instruction_durations.")
```
Comment (itoko) [inline qiskit/compiler/transpile.py:431]: Just note that these lines make impossible to transpile circuits with only delays. barriers, and cal gates for simulator backend, while `test_transpile_delay_circuit_without_scheduling_method_or_durs` succeeds without these lines. However, I cannot think of any use case of simulating such circuits and these lines help provide more informative error message. So adding these lines makes sense to me.

Comment (lcapelluto) [reply qiskit/compiler/transpile.py:431]: cal gates will populate durations, right? Is there a better way to check this? If I run `transpile(qc)` on a circuit that has delays, I get an error something like "Error: Duration for operation 'cx' on qubits [0, 1] was not found"

File: test/python/circuit/test_scheduled_circuit.py [hunk 5]
```diff
@@ -109,13 +109,6 @@ def test_cannot_schedule_circuit_with_mixed_SI_and_dt_when_no_one_tells_dt(self)
         with self.assertRaises(QiskitError):
             transpile(qc, self.backend_without_dt, scheduling_method='alap')
 
-    def test_can_execute_delay_circuit_when_schedule_circuit_off(self):
-        qc = QuantumCircuit(2)
-        qc.h(0)
-        qc.delay(500, 1)
-        qc.cx(0, 1)
-        execute(qc, backend=self.backend_with_dt, schedule_circuit=False)
```
Comment (lcapelluto) [inline test/python/circuit/test_scheduled_circuit.py:117]: I don't know how this wasn't failing before, but I also don't know why it hangs rather than fails. (It hangs on `job = backend.run(qobj, **run_config)` of `execute.py`)

Comment (itoko) [reply test/python/circuit/test_scheduled_circuit.py:117]: Yeah, this test must be removed. I confirmed it hangs in my local env. I don't know why it doesn't fail or how it wasn't failing before, either.

File: qiskit/compiler/transpile.py (hunk 4 above)
Comment (itoko) [reply qiskit/compiler/transpile.py:434]: I don't have any good idea; we may accept less informative error (no duration found), or, we may have to check if circuits are composed only by instructions which can populate durations (I think it's too much). I think we can leave these lines as you've added for now.

**Review by itoko — APPROVED** [review]
Thanks for your updates!

File: test/python/circuit/test_scheduled_circuit.py (hunk 5 above)
Comment (lcapelluto) [reply test/python/circuit/test_scheduled_circuit.py:117]: Thank you for checking it out!
